### ISSUE TITLE
Calling ZipZap's updateEntries method causes a crash in Swift

### ISSUE THREAD

**Issue by marulya** [pixelglow/zipzap#94]
I've been trying to create an archive using the ZipZap library (https://github.com/pixelglow/zipzap) and to be more specific the 8.0 release. The library is consumed in Swift code.

I've read that when calling:

    newArchive.updateEntries(archiveItems, error: error)

Requires the "archiveItems" to be NSMutableArray and it is defined as such. At this point I tried creating an archive that contains just one directory so the array looks like:

    var archiveItems = NSMutableArray()
    archiveItems.addObject(ZZArchiveEntry(directoryName: "\(archiveName)/"))

"newArchive" is created the following way:

    var newArchive = ZZArchive(URL: NSURL(fileURLWithPath: archivePath), error: error)

The error I see is: EXC_BREAKPOINT(code=1, subcode=0x1001bc998)

And I've also seen: EXC_BREAKPOINT(code=1, subcode=0x100100998)

In case I do not call the updateEntries method the code does not crash. So my assumption is that the crash happens inside this method.

Thanks for the help!

**Comment by pixelglow** [comment]
I don't see anything on inspection that would cause a crash. Please send me the minimum program source that triggers the problem, and possibly also a Objective-C version of that source if you can manage it.

As an aside, `updateEntries:error:` does not require an `NSMutableArray` -- an `NSArray` will do fine. We used `NSMutableArray` in the examples to illustrate updating the existing entries of the zip file. You can use an `NSArray` to store totally new entries (as in your code), or even make an `NSArray` out of some old entries and some new entries.

**Comment by marulya** [comment]
Hi,

Thank you for taking the time to look into this!

I've prepared a minimal project for you that replicates the crash. It is written in Swift because we are using Swift in our project and this is what we are interested in. In case the issue is related to Objective-C and Swift compatibility we could consider writing this particular piece in Objective-C.

When I was testing the project I saw a message that was not visible in the previous code:

    fatal error: unexpectedly found nil while unwrapping an Optional value

I've checked if I pass a nil value anywhere and I do not pass a nil value, but it turned out that the instance of ZZArchive that I've created is actually nil. So it makes perfect sense for the code to crash trying to access a method in this instance. Question is why is this instance nil.

So I've tried digging a bit further and checked if the NSURL instance that is passed to instance creation has a valid value in it but it seemed to be correct. It contained the following:

    file:///private/var/mobile/Containers/Data/Application/B42719C0-FFEB-4409-BC15-090606B3A4E1/tmp/archive.zip

Which is exactly what I would expect. And since this method has only 2 parameters the URL and the Error pointer I've also checked that passing an error pointer and a nil for error pointer does not make a difference.

At this point it seems very likely to me that I've messed something with the project setup. So here is a description of what I've done. The 8.0 release version that I've downloaded was built in Release and the resulting .a file and .h files were taken from the ~/Library/Developer/Xcode/DerivedData folder as described in the readme. Since we use Swift we have added the zipzap.h to the bridging header:

    #import "zipzap.h"

And we have also included the following dependencies:
* ImageIO.framework
* Foundation.framework
* libz.dylib

We are using Xcode 6.1.1
Pretty much this is what we've done.

Here is a link to the archive containing replicating project: https://www.dropbox.com/s/88ooy5yay5652sr/ZipZap-Swift-Crash.zip?dl=0

Hope this helps. In case I've missed any information that will be useful to you I'll do my best to provide it.

Once again thank you for looking into this!

**Comment by pixelglow** [comment]
The immediate fix is to use

    var newArchive = ZZArchive(URL: url, options:[ZZOpenOptionsCreateIfMissingKey: true], error: &archiveError)

You'll need to pass in `ZZOpenOptionsCreateIfMissingKey: true` as an option so that the initialiser creates the zip file if it doesn't exist. By default `ZZArchive` only opens existing zip files. This is based on the Unix `open` function needing a `O_CREAT` flag to create a file if it doesn't exist, and is similar to usage in LevelDB etc.

You would see the error message on initialization if you grab it from the error parameter instead of passing in a `NSErrorPointer`.

**Comment by marulya** [comment]
Thank you so much for having the time to look into this!

At the end indeed it was me messing things up. Apologies for wasting your time on that. If I have passed a proper error parameter probably would have saved you the time. And it a shame I did not see that this option is being passed in your "Writing a new zip file:" example.

Anyways, passing the options did work for me and we now have a working archiving, hooray.

Once again thank you, and sorry for wasting your time.

All the best!
